kubeflow/trainer/backends/kubernetes/backend.py

```diff
@@ -209,6 +209,35 @@ def train(
             trainer_overrides = spec_section.get("trainer", {})
             pod_template_overrides = spec_section.get("podTemplateOverrides")
 
+        # Auto-add progression tracking annotations for RHAI trainers
+        if (
+            trainer
+            and hasattr(trainer, "__module__")
+            and "rhai" in trainer.__module__
+            and hasattr(trainer, "enable_progression_tracking")
+            and trainer.enable_progression_tracking
+        ):
+            # Import here to avoid circular dependency
+            from kubeflow.trainer.rhai.constants import (
+                ANNOTATION_FRAMEWORK,
+                ANNOTATION_METRICS_POLL_INTERVAL,
+                ANNOTATION_METRICS_PORT,
+                ANNOTATION_PROGRESSION_TRACKING,
+            )
+
+            progression_annotations = {
+                ANNOTATION_PROGRESSION_TRACKING: "enabled",
+                ANNOTATION_METRICS_PORT: str(trainer.metrics_port),
+                ANNOTATION_METRICS_POLL_INTERVAL: str(trainer.metrics_poll_interval_seconds),
+                ANNOTATION_FRAMEWORK: "transformers",
+            }
+
+            # Merge with existing spec_annotations
+            if spec_annotations is None:
+                spec_annotations = progression_annotations
+            else:
+                spec_annotations = {**spec_annotations, **progression_annotations}
```

Comment:
suggestion: can we move this logic into `rhai/utils.py` to be the same as `get_trainer_crd_from_transformers_trainer`?

Ideally we want the changes in upstream files to be as minimal as possible.

How about tidying the if clause a little?

```suggestion
        if isinstance(trainer, RHAITrainer):
            rhai_annotations = rhai_utils.get_rhai_train_job_annotations(trainer)
            if spec_annotations is None:
                spec_annotations = rhai_annotations
            else:
                # apply spec_annotations last to allow users to override the rhai annotations 
                spec_annotations = rhai_annotations | spec_annotations
```

Reply:
Did you mean to apply another change here? This is still a big blob of extra code in the upstream `backend.py` file which is likely to cause bigger merge conflicts.